File: zgadywanko.scala
import collection.mutable.Set
import io.StdIn.readLine
import util.Random

def getYesNo(prompt: String): Boolean =
  readLine(s"${prompt} (y/n) ").strip.toLowerCase match
    case "y" => true
    case "n" => false
    case _ => getYesNo(prompt)

def randBigInt(a: BigInt, b: BigInt): BigInt =
  val result = BigInt((b - a).bitLength, Random)
  if result <= b - a then
    result + a
  else
    randBigInt(a, b)

def getNum(prompt: String): BigInt =
  try
    BigInt(readLine(s"${prompt} ").strip)
  catch case _: NumberFormatException =>
    getNum(prompt)

@main
def main(args: String*) =
  println("Welcome to The Miraculous Game of Zgadywanko!")

  require(args.size <= 2, "Too many arguments")
  val maxNum = args.lift(0).fold(BigInt(100))(BigInt.apply)
  val numc = args.lift(1).fold(1)(_.toInt)
  require(numc >= 1, "The number of numbers must not be less than 1")
  require(maxNum >= numc, s"The upper bound must not be less than ${numc}")

  var shouldContinue = true
  while shouldContinue do
    println("Guess the " + (if numc == 1 then "number" else s"${numc} numbers") + s" from 1 to ${maxNum}.")
    val nums = Set.from(Iterator.continually(randBigInt(BigInt(1), maxNum)).distinct.take(numc))

    while nums.nonEmpty do
      val guess = getNum("Your guess:")

      val didFindAny = nums.remove(guess)
      if nums.isEmpty then
        println("You've found the last number, congratulations!")
      else
        if didFindAny then
          println("That's one of the numbers, good job!")

        val lessc = nums.count(_ < guess)
        val greaterc = nums.size - lessc
        print(if didFindAny then "Still, too " else "Too ")
        if lessc > 0 then
          print(if lessc == 1 then "high" else s"high for ${lessc}")
        if lessc > 0 && greaterc > 0 then
          print(" and too ")
        if greaterc > 0 then
          print(if greaterc == 1 then "low" else s"low for ${greaterc}")
        println("!")

    shouldContinue = getYesNo("Do you want to play again?")

  println("Thanks for playing, bye!")

File: zgadywanko.rb
require 'set'

def get_yes_no(prompt)
  print "#{prompt} (y/n) "
  case $stdin.readline.strip.downcase
    when 'y' then true
    when 'n' then false
    else get_yes_no prompt
  end
end

puts 'Welcome to The Miraculous Game of Zgadywanko!'

raise 'Too many arguments' if $*.length > 2
max_num = $*[0]&.to_i || 100
numc = $*[1]&.to_i || 1
raise 'The number of numbers must not be less than 1' if numc < 1
raise "The upper bound must not be less than #{numc}" if max_num < numc

while true
  puts 'Guess the ' + (numc == 1 ? 'number' : "#{numc} numbers") + " from 1 to #{max_num}."
  nums = Enumerator.produce { rand(1 .. max_num) }.lazy.uniq.take(numc).to_set

  while true
    begin
      print 'Your guess: '
      guess = Integer $stdin.readline.strip
    rescue ArgumentError
      retry
    end

    did_find_any = nums.delete?(guess)
    if nums.empty?
      puts "You've found the last number, congratulations!"
      break
    end
    puts "That's one of the numbers, good job!" if did_find_any

    lessc = nums.count { |x| x < guess }
    greaterc = nums.length - lessc
    print(did_find_any ? 'Still, too ' : 'Too ')
    print(lessc == 1 ? 'high' : "high for #{lessc}") if lessc > 0
    print ' and too ' if lessc > 0 and greaterc > 0
    print(greaterc == 1 ? 'low' : "low for #{greaterc}") if greaterc > 0
    puts '!'
  end

  break unless get_yes_no 'Do you want to play again?'
end

puts 'Thanks for playing, bye!'
